website/src/pages/tech-specs.md

```diff
@@ -0,0 +1,350 @@
+
+
+
+## Transaction Log (Timeline)
+
+Data consistency in Hudi is provided using Multi-version Concurrency Control (MVCC). Every transactional action on the Hudi table creates a new entry (instant) in the timeline. All transactional actions follows the state transition below
+
+* **requested** - Action is requested to start on the timeline
+* **inflight** - Action has started running and is currently in-flight
+* **completed** - Action has completed running
+
+All actions and the state transitions are registered with the timeline using an atomic put of special meta-file inside the  *.hoodie* directory. The requirement from the underlying storage system is to support an atomic-put and read-after-write consistency. The meta file structure is as follows
+
+```
+[Action timestamp].[Action type].[Action state] 
+```
+
+* **Action timestamp** - Monotonically increasing value to denote strict ordering of actions in the timeline. This could be provided by an external token provider or rely on the system epoch time at millisecond granularity. 
+
+* **Action type** - Type of action. The following are the possition actions on the hudi timeline. 
+
+* | Action type   | Description                                                  |
+  | ------------- | ------------------------------------------------------------ |
+  | commit        | Commit denotes an **atomic write (inserts, updates and deletes)** of records in a table. A commit in Hudi is an atomic way of updating data, metadata and indexes. The guarantee is that all or none the changes within a commit will be visible to the readers |
+  | deltacommit   | Special version of `commit` which is applicable only on a Merge-on-Read storage engine. The writes are accumulated and batched to improve write performance |
+  | rollback      | Rollback denotes that the changes made by the corresponding commit/delta commit was unsuccessful & hence rolled back, removing any partial files produced during such a write |
+  | savepoint     | Savepoint is a special marker to ensure a particular commit is not automatically cleaned. It helps restore the table to a point on the timeline, in case of disaster/data recovery scenarios |
+  | restore       | Restore denotes that a particular Savepoint was restored     |
+  | clean         | Maintenance activity that cleans up versions of data files that no longer will be accessed |
+  | compaction    | Maintenance to optimize the storage for query performance. This action applies the batched up updates from `deltacommit` and re-optimizes data files for query performance |
+  | replacecommit | Maintenance activity to cluster the data for better query performance. This action is different from a `commit` in that the the table state before and after are logically equivalent |
+  | indexing      | [todo]                                                       |
+  | schemacommit  | [todo]                                                       |
```

Comment:
This action denotes persistence of merged schema under `.hoodie/schema`. It only happens when full schema evolution feature is enabled. Since that feature is still experimental, it is disabled by default. This makes hoodie metadata like a schema registry as well. So, we have full history schema evolution. In the future, plan is to move this inside `.hoodie/metadata`. @xiarixiaoyao can add more details.